x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java

```diff
@@ -937,4 +941,125 @@ private LogicalPlan rule(Eval eval) {
         }
 
     }
+
+    /**
+     * Normalize aggregation functions by:
+     * 1. replaces reference to field attributes with their source
+     * 2. in case of Count, aligns the various forms (Count(1), Count(0), Count(), Count(*)) to Count(*)
+     */
+    // TODO waiting on https://github.com/elastic/elasticsearch/issues/100634
+    static class NormalizeAggregate extends Rule<LogicalPlan, LogicalPlan> {
+
+        @Override
+        public LogicalPlan apply(LogicalPlan plan) {
+            AttributeMap<Expression> aliases = new AttributeMap<>();
+
+            // traverse the tree bottom-up
+            // 1. if it's Aggregate, normalize the aggregates
+            // regardless, collect the attributes but only if they refer to an attribute or literal
+            plan = plan.transformUp(p -> {
+                if (p instanceof Aggregate agg) {
+                    p = normalize(agg, aliases);
+                }
+                p.forEachExpression(Alias.class, a -> {
+                    var child = a.child();
+                    if (child.foldable() || child instanceof NamedExpression) {
+                        aliases.putIfAbsent(a.toAttribute(), child);
+                    }
+                });
+
+                return p;
+            });
+            return plan;
+        }
+
+        private static LogicalPlan normalize(Aggregate aggregate, AttributeMap<Expression> aliases) {
+            var aggs = aggregate.aggregates();
+            List<NamedExpression> newAggs = new ArrayList<>(aggs.size());
+            boolean changed = false;
+
+            for (NamedExpression agg : aggs) {
+                if (agg instanceof Alias as && as.child() instanceof AggregateFunction af) {
```

Comment:
Here and below we assume that the agg is aliased. Is this always correct? I'm thinking of queries like
```
...
| eval x = 1, y = x
| stats min(x), min(y)
```
in this case the introduced stats are aliased to `"min(x)"` and `"min(y)"`, making this code correct, right?

Reply:
I'll answer the question with a question - no need to answer here, where does this potential aliasing take place?

Reply:
Ok, looks like the logical plan builder visits the aggregation expressions, where `visitFields` retrns a list of aliases.

That still doesn't fully answer the question, though - even though initially the aggregation expressions can only be a list of aliases, they could end up being a different kind of `NamedExpression` because of some optimization rule or so. There's nothing inside `Aggregates.java` that enforces that we always have aliases.

IMHO, if there's a contract that `Aggregate::aggregates` always returns a list of aliases, this contract should be spelled out explicitly in a javadoc for `Aggregate` or, better yet, it should be enforced at compile time by changing the return type of `Aggregate::aggregates` (and, accordingly, the signature of `Aggregate`'s constructor).